### ISSUE TITLE
Serious regression of warp!

### ISSUE THREAD

**Issue by jianghaizhu** [JuliaImages/ImageTransformations.jl#60]
Some serious memory allocation and regression in Julia 0.7. Here is a MWE.

```
julia> img = rand(100, 100);
julia> tfm = recenter(RotMatrix(-pi/4), center(img));
```
In Julia 0.6,
```
julia> @btime warp(img, tfm);
  422.876 μs (5 allocations: 157.72 KiB)
```
In Julia 0.7,
```
julia> @btime warp(img, tfm);
  8.478 ms (261626 allocations: 9.07 MiB)
```
Looking at the code of `warp!`,
```
function warp!(out, img::AbstractExtrapolation, tform)
    @inbounds for I in CartesianIndices(axes(out))
        out[I] = _getindex(img, tform(SVector(I.I)))
    end
    out       <==== should not return anything
end
```
Obviously, `warp!` should not return `out`. But eliminating the return only helps marginally. But I cannot dig any further. Might be some problem with `_getindex`.

**Comment by kmsquire** [comment]
Returning `out` is not a problem. It doesn't allocate any additional memory, and allows the result to be used pipelines. (Unlike, e.g., pandas, there's no convention that functions which modify their parameters in place return nothing.)

**Comment by jianghaizhu** [comment]
Yes. Returning is not a problem. The question here is what the problem is.

**Comment by RalphAS** [comment]
Profiling points to `getindex`, which we hoped was fixed by https://github.com/JuliaLang/julia/pull/28707. But the OP's example is even worse on a recent nightly of Julia (after that PR).  
Is @timholy's overhaul of Interpolations.jl likely to make compiler optimization any more effective for interpolation objects, or do we need to pray for more magic from Keno?

**Comment by RalphAS** [comment]
I'm glad that @Evizero caught my mistake. The sad thing is that [I knew better last year](https://discourse.julialang.org/t/concerns-about-splatting/6664/6).

Is it the case that splatting (almost?) anything other than a Tuple leads to dynamic dispatch, which can foul up inference in various (hard-to-predict) ways?

**Comment by Evizero** [comment]
No worries, these things happen to us all. If anything I should have caught it in the review. Your contributions are a clear net positive and we are lucky to have you. The occasional regression is to be expected and nothing we can't fix
